File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        
        temp = head

        while n > 0:
            temp = temp.next
            n -= 1

        if not temp:
            return head.next    
        
        start = head

        while temp.next:
            temp = temp.next
            start = start.next

        start.next = start.next.next    

        return head           

File: Data Structures & Algorithms/reorder-linked-list/submission-3.java
// class ListNode {
//     int val;
//     ListNode next;
//     ListNode(int val) { this.val = val; }
// }

class Solution {

    static ListNode left = null;
 
    public void reorderList(ListNode head)
    {
        if(head == null || head.next == null) return;   
        left = head;
        rearrangeH(left, null);
    }
    
    public static void rearrangeH(ListNode curr, ListNode prev)
    {
        if(curr.next != null) rearrangeH(curr.next, curr);
        if(left != null && left.next != null) {
            prev.next = null;
            ListNode saveNext = left.next;
            left.next = curr;
            curr.next = saveNext;
            left = saveNext;
        }
    }
}
